libvast/vast/format/json.hpp

```diff
@@ -41,4 +53,202 @@ class writer : public printer_writer<event_printer>{
   }
 };
 
+struct default_selector {
+  caf::optional<vast::record_type> operator()(const vast::json::object&) {
+    return t;
+  }
+
+  caf::error schema(vast::schema sch) {
+    auto entry = *sch.begin();
+    if (!caf::holds_alternative<vast::record_type>(entry))
+      return make_error(vast::ec::invalid_configuration,
+                        "only record_types supported for json schema");
+    t = flatten(caf::get<vast::record_type>(entry));
+    return caf::none;
+  }
+
+  vast::schema schema() const {
+    vast::schema result;
+    if (t)
+      result.add(*t);
+    return result;
+  }
+
+  static const char* name() {
+    return "json-reader";
+  }
+
+  caf::optional<vast::record_type> t = caf::none;
+};
+
+template <class Selector = default_selector>
+class reader final : public multi_layout_reader {
+public:
+  using super = multi_layout_reader;
+
+  /// Constructs a json reader.
+  /// @param input The stream of logs to read.
+  explicit reader(caf::atom_value table_slice_type,
+                  std::unique_ptr<std::istream> in = nullptr);
+
+  void reset(std::unique_ptr<std::istream> in);
+
+  caf::error schema(vast::schema sch) override;
+
+  caf::error schema(vast::type, vast::schema);
+
+  vast::schema schema() const override;
+
+  const char* name() const override;
+
+protected:
+  caf::error read_impl(size_t max_events, size_t max_slice_size,
+                       consumer& f) override;
+
+private:
+  using iterator_type = std::string_view::const_iterator;
+
+  // void patch(std::vector<data>& xs);
+
+  Selector selector;
```

Comment:
```suggestion
  Selector selector_;
```

Plus all the changes that would go with this.